pennylane/qnodes/base.py

```diff
@@ -131,6 +131,96 @@ def decompose_queue(ops, device):
     return new_ops
 
 
+def _compare_objects(object1, object2):
+    """Checks if two objects are equal.
+
+    Supports equality checks for NumPy arrays and TensorFlow/PyTorch tensors without requiring an
+    import of PyTorch or TensorFlow.
+
+    The function uses a broad try/except, which returns False on the exception. This allows us to
+    say that two incompatible objects are not equal without throwing an exception.
+
+    Args:
+        object1: first item to be compared
+        object2: second item to be compared
+
+    Returns:
+        bool: whether the two items are equal
+    """
+    try:
+        if type(object1) != type(object2):
+            return False
+
+        comparison = object1 == object2
+
+        if isinstance(comparison, bool):
+            return comparison
+
+        try:
+            return comparison.all()
+        except AttributeError:
+            return comparison.numpy().all()
+    except Exception:
+        return False
+
+
+def _compare_lists(list1, list2):
+    """Checks if two lists are equal.
+
+    Supports nested lists that can include NumPy arrays and TensorFlow/PyTorch tensors without
+    requiring an import of PyTorch or TensorFlow.
+
+    Args:
+        list1 (list): first list to be compared
+        list2 (list): second list to be compared
+
+    Returns:
+        bool: whether the two lists are equal
+    """
+    list1 = _flatten(list1)
+    list2 = _flatten(list2)
+
+    for a1, a2 in zip(list1, list2):
+        if not _compare_objects(a1, a2):
+            return False
+
+    return True
+
+
+def _compare_dicts(dict1, dict2):
+    """Checks if two dictionaries are equal.
+
+    Supports dictionaries with values that can be nested lists that may include NumPy arrays and
+    TensorFlow/PyTorch tensors without requiring an import of PyTorch or TensorFlow.
+
+    Args:
+        dict1 (dict): first dictionary to be compared
+        dict2 (dict): second dictionary to be compared
+
+    Returns:
+        bool: whether the two dictionaries are equal
+    """
+    if dict1 is None:
+        dict1 = {}
+    if dict2 is None:
+        dict2 = {}
+
+    if len(dict1) != len(dict2):
+        return False
+
+    values_1 = []
+    values_2 = []
+    for key, value in dict1.items():
+        values_1.append(value)
+        v2 = values_2.get(key)
+        if v2:
+            values_2.append(v2)
+        else:
+            return False
+
+    return _compare_lists(values_1, values_2)
```

Comment:
This was the way I thought of to compare args and kwargs. What do you think @josh146 @antalszava?